hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/hudi/TestHoodieSparkSqlWriter.scala

```diff
@@ -1179,6 +1177,88 @@ class TestHoodieSparkSqlWriter {
+      DataSourceWriteOptions.PARTITIONPATH_FIELD.key() -> "partition:simple",
+      HoodieWriteConfig.MERGE_ALLOW_DUPLICATE_ON_INSERTS_ENABLE.key() -> "false")
+
+    try {
+      // verify exception is thrown when HoodieWriteConfig.COMBINE_BEFORE_INSERT is enabled
+      val tmpDF = Seq((1, "a1", 10, 1000, "2021-10-16")).toDF("id", "name", "value", "ts", "dt")
+      tmpDF.write.format("org.apache.hudi")
+        .options(initialOpts + (HoodieWriteConfig.COMBINE_BEFORE_INSERT.key() -> "true"))
+        .mode(SaveMode.Append)
+        .save(tempBasePath)
+    } catch {
+      case e: HoodieException =>
+        assertTrue(e.getMessage.contains(s"${classOf[KeylessKeyGenerator].getName} can not be used with config enabled " +
+          s"{${HoodieWriteConfig.COMBINE_BEFORE_INSERT.key()}}"), e.getMessage)
+    }
+  }
+
+  @ParameterizedTest
+  @ValueSource(strings = Array("COPY_ON_WRITE", "MERGE_ON_READ"))
+  def testAutoGenerationOfRecordKeys(tableType: String): Unit = {
+
+    val dataGen = new HoodieTestDataGenerator()
+    val initialOpts = Map(
+      HoodieWriteConfig.TBL_NAME.key -> hoodieFooTableName,
+      "hoodie.insert.shuffle.parallelism" -> "1",
+      "hoodie.upsert.shuffle.parallelism" -> "1",
+      DataSourceWriteOptions.TABLE_TYPE.key -> tableType,
+      DataSourceWriteOptions.KEYGENERATOR_CLASS_NAME.key() -> "org.apache.hudi.keygen.KeylessKeyGenerator",
+      DataSourceWriteOptions.OPERATION.key() -> UPSERT_OPERATION_OPT_VAL,
+      DataSourceWriteOptions.PARTITIONPATH_FIELD.key() -> "partition:simple",
+      HoodieWriteConfig.MERGE_ALLOW_DUPLICATE_ON_INSERTS_ENABLE.key() -> "false")
+
+    var totalRecs = 0
+    val extraOptsList = List(
+      Map.empty,
+      Map(
+        DataSourceWriteOptions.PRECOMBINE_FIELD.key -> "timestamp",
+        DataSourceWriteOptions.RECORDKEY_FIELD.key -> "_row_key")
+    )
+    for (extraOpt <- extraOptsList) {
+      val opts = initialOpts ++ extraOpt
+      for (x <- 1 to 5) {
+        val instantTime = "00" + x
+        val genRecsList = if (x == 1) {
+          totalRecs += 100 * 2
+          val inserts = dataGen.generateInserts(instantTime, 100)
+          inserts.addAll(inserts)
+          inserts
+        } else {
+          totalRecs += 10
+          dataGen.generateUniqueUpdates(instantTime, 10)
+        }
+        val records = recordsToStrings(genRecsList).toList
+        val inputDF = spark.read.json(spark.sparkContext.parallelize(records, 2))
+        inputDF.write.format("org.apache.hudi")
+          .options(opts)
+          .mode(SaveMode.Append)
+          .save(tempBasePath)
+
+        val snapshotDF = spark.read.format("org.apache.hudi")
+          .load(tempBasePath)
+        assertEquals(totalRecs, snapshotDF.count())
+      }
+    }
+
+    val metaClient = HoodieTableMetaClient.builder().setConf(spark.sparkContext.hadoopConfiguration).setBasePath(tempBasePath).build()
+    assertEquals(metaClient.getTableConfig.getTableType, HoodieTableType.COPY_ON_WRITE)
```

Comment:
can we write tests for exception/non happy paths. 
1. if combine.before.insert is set, and try to send duplicates records, all duplicates should be present in snapshot read. 
2. verify that preCombine field is not set in tableConfig. ensure tests sets it explicitly.
3. if record key field config is set, it should not throw if keyless is enabled. and try to inject records w/ same record keys.